File: solutions/java/circular-buffer/2/src/main/java/CircularBuffer.java
import java.util.Arrays;

public class CircularBuffer<T> {

    private final Object[] buffer;
    private int head = 0;
    private int tail = 0;
    private int count = 0;

    public CircularBuffer(int capacity) {
        this.buffer = new Object[capacity];
    }

    public T read() throws BufferIOException {
        if (isEmpty()) {
            throw new BufferIOException("Tried to read from empty buffer");
        }

        @SuppressWarnings("unchecked")
        T value = (T) buffer[head];

        buffer[head] = null;
        head = (head + 1) % buffer.length;
        count--;

        return value;
    }

    public void write(T data) throws BufferIOException {
        if (isFull()) {
            throw new BufferIOException("Tried to write to full buffer");
        }

        buffer[tail] = data;
        tail = (tail + 1) % buffer.length;
        count++;
    }

    public void overwrite(T data) {
        if (isFull()) {
            buffer[head] = data;
            head = (head + 1) % buffer.length;
            tail = (tail + 1) % buffer.length; 
        } else {
            try {
                write(data);
            } catch (BufferIOException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public void clear() {
        Arrays.fill(buffer, null);
        head = 0;
        tail = 0;
        count = 0;
    }

    private boolean isEmpty() {
        return count == 0;
    }

    private boolean isFull() {
        return count == buffer.length;
    }
}

File: solutions/python/circular-buffer/2/circular_buffer.py
class BufferFullException(BufferError):
    def __init__(self, message):
        super().__init__(message)


class BufferEmptyException(BufferError):
    def __init__(self, message):
        super().__init__(message)


from collections import deque

class CircularBuffer:
    def __init__(self, capacity):
        self.buffer = deque(maxlen=capacity)

    def write(self, data):
        if len(self.buffer) == self.buffer.maxlen:
            raise BufferFullException("Circular buffer is full")
        self.buffer.append(data)

    def read(self):
        if not self.buffer:
            raise BufferEmptyException("Circular buffer is empty")
        return self.buffer.popleft()

    def overwrite(self, data):
        self.buffer.append(data)

    def clear(self):
        self.buffer.clear()